### PR TITLE
[SPARK-12636][SQL] Update UnsafeRowParquetRecordReader to support reading files directly.

### PR DESCRIPTION
**Pull request by nongli**
As noted in the code, this change is to make this component easier to test in isolation.

### CODE REVIEW

**Comment by rxin** [comment]
The high level change looks good to me, although I have to admit I'm not familiar with details of Parquet.

cc @nongli and @liancheng

**Comment by nongli** [comment]
This doesn't really do much but just to make this component create-able without the hadoop machinery.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java [hunk 1]
```diff
@@ -125,20 +129,80 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
...
     for (BlockMetaData block : blocks) {
       this.totalRowCount += block.getRowCount();
     }
   }
 
+  /**
+   * Returns the list of files at 'path' recursively. This skips files that are ignored normally
+   * by MapReduce.
+   */
+  public static List<String> listDirectory(File path) throws IOException {
```
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java]: Is this only used by tests?

**Comment by davies** [comment]
LGTM, except one minor comment.

**Comment by davies** [comment]
Merging this to unblock the following up PR (could be addressed there).

**Pull request closed without merging**